feat(api): nvim_win_set_config can move floatwin to another tabpage

glepnir opened the pull request:
Problem: nvim_win_set_config cannot move floating windows to different tab pages
Solution: Add cross-tab window movement logic with proper insertion point handling 

Fix #33790 

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

glepnir committed:
feat(api): nvim_win_set_config can move floatwin to another tabpage

Problem: nvim_win_set_config can't move floating windows to different tab pages.

Solution: allow it.

Co-authored-by: Sean Dewar <6256228+seandewar@users.noreply.github.com>

glepnir commented:
cc @seandewar

seandewar commented on src/nvim/api/win_config.c:
```diff
@@ -639,6 +656,51 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+
+      // Insert into target tabpage
+      win_append(target_after, win, parent_tp == curtab ? NULL : parent_tp);
+      // Ensure target tabpage has valid curwin
+      if (parent_tp != curtab && parent_tp->tp_curwin == NULL) {
+        parent_tp->tp_curwin = parent_tp->tp_firstwin;
+      }
```
Not sure what this is checking. Why would `parent_tp->tp_curwin` be NULL here?

You _should_ however check that `win_tp->tp_curwin` is set to something valid if it was previously equal to `win`, like is done for splits.

seandewar reviewed (commented):
I don't have a lot of time to review, so here's some quick comments:

glepnir commented on src/nvim/api/win_config.c:
```diff
@@ -639,6 +656,51 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+      }
+
+      if (win == curwin) {
+        win_goto(win_float_find_altwin(win, NULL));
```
cool update for validate

seandewar reviewed (commented):
Latest changes seem to be overcomplicating things.

Check out what is done for splits for inspiration; many of the same checks are relevant.

glepnir commented:
cleaned up some redundant logic and reorganized the implementation

seandewar commented on src/nvim/api/win_config.c:
```diff
@@ -639,6 +656,63 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+        }
+
+        win_append(target_after, win, parent_tp == curtab ? NULL : parent_tp);
+        if (win_tp->tp_curwin == win) {
+          win_tp->tp_curwin = altwin ? altwin : win_float_find_altwin(win, win_tp);
+        }
```
```suggestion
        // If `win` was the curwin of its old tabpage, select a new curwin for it.
        if (win_tp->tp_curwin == win) {
          win_tp->tp_curwin = win_float_find_altwin(win, win_tp);
        }
```

Don't reuse `altwin`; `win_goto` autocommands may have freed it.

seandewar commented on src/nvim/api/win_config.c:
```diff
@@ -639,6 +656,63 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+      if (!can_move(win, win_tp != parent_tp, err)) {
+        return;
+      }
+      win_T *altwin = NULL;
+
+      if (win == curwin) {
+        altwin = win_float_find_altwin(win, NULL);
+        win_goto(altwin);
+
+        if (curwin == win) {
+          api_set_error(err, kErrorTypeException, "Failed to switch away from window %d",
+                        win->handle);
+          return;
+        }
+
+        if (!win_valid_any_tab(win)) {
+          api_set_error(err, kErrorTypeException, "Window was closed by autocommands %d",
+                        win->handle);
+          return;
+        }
+
+        win_tp = win_find_tabpage(win);
+        parent_tp = win_find_tabpage(parent);
+      }
```
```suggestion
      if (win == curwin) {
        win_goto(win_float_find_altwin(win, NULL));
        
        // Autocommands may have been a real nuisance and messed things up...
        if (curwin == win) {
          api_set_error(err, kErrorTypeException, "Failed to switch away from window %d",
                        win->handle);
          return;
        }
        win_tp = win_find_tabpage(win);
        parent_tp = win_find_tabpage(parent);
        if (!win_tp || !parent_tp) {
          api_set_error(err, kErrorTypeException, "Target windows were closed");
          goto restore_curwin;
        }
        if (!win->w_floating) {
          api_set_error(err, kErrorTypeException, "Window %d was made non-floating",
                        win->handle);
          goto restore_curwin;
        }
        if (!can_move(win, win_tp != parent_tp, err)) {
          goto restore_curwin;
        }
```
Implement `restore_curwin` so it attempts to `win_goto` the old `curwin` and return.

The checks are pretty similar to what's needed for splits; maybe some of this can be de-duplicated if it doesn't make things look worse?

Also the `can_move` check has to be repeated like I've done here; autocommands may have made the window external via `<C-W>ge` or similar, for example. (May be a bug that the repeated check was missing for splits...? :melting_face:)

glepnir replied:
I will clean up some redundant logic later.

seandewar commented on src/nvim/api/win_config.c:
```diff
@@ -639,6 +656,63 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+          win_tp->tp_curwin = altwin ? altwin : win_float_find_altwin(win, win_tp);
+        }
+
+        if (parent_tp != curtab) {
+          if (ui_has(kUIMultigrid)) {
+            ui_call_win_hide(win->w_grid_alloc.handle);
+          }
+          ui_comp_remove_grid(&win->w_grid_alloc);
+        } else {
+          win->w_hl_needs_update = true;
+        }
+        win->w_viewport_invalid = true;
```
I dunno if this is correct. Someone else ought to check.

glepnir replied:
np

seandewar commented on src/nvim/api/win_config.c:
```diff
@@ -639,19 +650,81 @@ void nvim_win_set_config(Window window, Dict(win_config) *config, Error *err)
+restore_curwin:
+  // If `win` was the original curwin, and autocommands didn't move it outside of curtab, be a
+  // good citizen and try to return to it.
+  if (curwin_moving_tp && win_valid(win)) {
```
From glancing at this, doesn't look like `curwin_moving_tp` is ever set when moving a float? (Add a test for this too)

glepnir replied:
missing set it before win_goto..

glepnir replied:
```
        // If `win` was the curwin of its old tabpage, select a new curwin for it.
        if (win_tp->tp_curwin == win) {
          win_tp->tp_curwin = win_float_find_altwin(win, win_tp);
        }
```
there may not correct ? I got a crash with this test

```lua
    it('sets curwin_moving_tp when moving float between tabs fails #test', function()
      local buf = api.nvim_create_buf(false, true)
      local float_win = api.nvim_open_win(buf, true, {
        relative = 'editor',
        row = 1,
        col = 1,
        width = 10,
        height = 5
      })
      command('tabnew')
      local tab2_win = api.nvim_get_current_win()
      command('tabprev')
      api.nvim_set_current_win(float_win)
      command('autocmd WinLeave * call nvim_win_close('..tab2_win..', true)')

      eq('', pcall_err(api.nvim_win_set_config, float_win, {
        relative = 'win',
        win = tab2_win,
        row = 0,
        col = 0,
        width = 10,
        height = 5
      }))

      eq(float_win, api.nvim_get_current_win())
    end)
    
    Assertion failed: (tp != curtab), function win_float_find_altwin, file winfloat.c, line 369.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==52902==ERROR: AddressSanitizer: ABRT on unknown address 0x0001963425b0 (pc 0x0001963425b0 bp 0x00016d609010 sp 0x00016d608ff0 T0)
    #0 0x0001963425b0 in pthread_kill+0x8 (libsystem_kernel.dylib:arm64+0x95b0)
    #1 0x000196282804 in abort+0x78 (libsystem_c.dylib:arm64+0x79804)
    #2 0x000196281a38 in assert_rtn+0x118 (libsystem_c.dylib:arm64+0x78a38)
    #3 0x00010422ecdc in win_float_find_altwin winfloat.c:369
    #4 0x0001029d19f4 in nvim_win_set_config win_config.c:698
    #5 0x000102946868 in handle_nvim_win_set_config dispatch_wrappers.generated.h:10035
    #6 0x0001036770cc in request_event channel.c:364
    #7 0x000103cbb43c in state_handle_k_event state.c:118
    #8 0x0001036f6dd8 in nv_event normal.c:6644
    #9 0x0001036b9a50 in normal_execute normal.c:1245
    #10 0x000103cbb184 in state_enter state.c:100
    #11 0x000103698ca4 in normal_enter normal.c:522
    #12 0x0001033c4890 in main main.c:655
```

zeertzjq commented on src/nvim/api/win_config.c:
```diff
@@ -380,6 +381,24 @@ static int win_split_flags(WinSplit split, bool toplevel)
+    return false;
+  }
+  if (wp->w_floating && wp->w_config.external) {
+    api_set_error(err, kErrorTypeException, "%s", "can not move external floating window");
```
```suggestion
    api_set_error(err, kErrorTypeException, "%s", "Cannot move external floating window");
```

justinmk reviewed (approved):
needs rebase, and approval from @seandewar

seandewar committed:
refactor(api): cleanup, more comments, more tests, news

- Factor out logic to keep nvim_win_set_config clean.
- Clean up a few things, remove redundant logic, reflow some lines.
- Add some more comments where appropriate.
- Don't consider negative "win", as that's only relevant for splits.
- Add more test coverage.
- Add news.txt entry.

seandewar committed:
fix(api): win_config_float_tp grid removal, redraw

Problem: when nvim_win_set_config moves a floatwin between tabpages, its grid
may remain if temporarily inside another tabpage. Also, things like tablines
aren't redrawn.

Solution: always remove its grid. Set must_redraw so things are redrawn even if
w_redr_type was already set for the old tabpage.

I don't think it's necessary to do anything extra here when removing the grid:

- win_ui_flush calls ui_call_win_hide anyway, and calling it manually ends up
  sending two win_hide events.
- ui_comp_remove_grid safely does nothing if the grid doesn't exist.
- w_pos_changed is set by win_config_float later, if that's needed. I think the
  pending_comp_index_update set by ui_comp_remove_grid is enough anyway, at
  least for making sure win_ui_flush sends win_hide.

Added test fails with the prior approach of checking `parent_tp != curtab`, but
also `win_tp == curtab`. (which is a better, but still flawed alternative)

The added redrawing here also supersedes setting w_hl_needs_update, and also
redraws stuff like the tabline to pass the new test.

seandewar committed:
refactor(window): lastwin_nofloating takes tp

seandewar committed:
fix(api): relax config validation for "win"

Problem: only possible to move floats between tabpages if relative=win, which
has the restrictive effect of also anchoring it to the target window.

Solution: allow "win" without "relative" or "split"/"vertical". Only assume
missing "win" is 0 if relative=win is given to maintain that behaviour. (or when
configuring a new window)

Also add an error when attempting to change a split into a float that's in
another tabpage, as this isn't actually supported yet. (until the next commit)

Maybe this could do with some bikeshedding. Unclear if "win" should require
"relative" to be given, like with "row"/"col"; this can be annoying though as
specifying "relative" requires other fields to be given too.

seandewar committed:
fix(api): don't config split as floatwin relative to itself

Problem: possible to configure a split as a floatwin with relative=win that is
relative to itself.

Solution: fix the check.

Not caused by this PR; just something I noticed when about to fix the validation
logic.

seandewar committed:
feat(api): nvim_win_set_config can move split to other tp as floatwin

Problem: not possible for nvim_win_set_config to convert a split to a floatwin,
then move it to another tabpage in one call.

Solution: allow it.

seandewar committed:
fix(winfloat): last_status when changing split to floatwin

Problem: converting a split to a floatwin may not remove the last statusline
when needed. (e.g: 'ls' is 1)

Solution: call last_status/win_comp_pos in win_new_float, after win_remove.

Also fix float_pos formatting for screen snapshots so it doesn't give a nil
error for external windows.

Not an issue from this PR.

seandewar committed:
fix(api): disallow moving window between tabpages in more cases

Problem: more cases where it may not be safe to move a window between tabpages.

Solution: check them.

Rather speculative... I haven't spend much time looking, but I didn't find
existing code that sets these locks to skip checking win_valid. (what I did find
called it anyway, like in win_close) Still, I think it's a good precaution for
what future code might do.

If the fact that nvim_win_set_config *actually* moves windows between tabpages
causes unforeseen issues, "faking" it like ":wincmd T" may be an alternative:
split a new window, close the old one, but instead also block autocmds, copy the
old window's config, and give it its handle?

seandewar committed:
fix(api): improve external window validation

Problem: "win" is allowed in external window configs in some cases. External
window converted to normal float can't move tabpages in one nvim_win_set_config
call. External window can't be turned into a normal split.

Solution: disallow setting "win" for external windows. Allow external window to
move tabpages, which turns it non-external. Allow external window to be turned
into a (non-external) split.

parse_win_config has more validation issues from not considering the window's
existing config enough (not from this PR). For example, zindex can be set for an
existing split if "split"/"vertical" isn't given, despite intending for that to
be an error. Plus the logic is confusing.

It could do with a refactor at some point...

justinmk commented on src/nvim/api/win_config.c:
```diff
@@ -608,7 +639,97 @@ static bool win_config_split(win_T *win, Dict(win_config) *config, WinConfig *fc
-#undef HAS_KEY_X
+}
+
+static bool win_config_float_tp(win_T *win, const Dict(win_config) *config,
```
brief docstring that explains the key idea of this?

justinmk reviewed (approved):
This is wonderful, thank you @seandewar + @glepnir

seandewar commented on test/functional/ui/float_spec.lua:
```diff
@@ -3824,6 +3821,18 @@ describe('float window', function()
+        )
+        api.nvim_open_win(buf, true, { external = true, width = 10, height = 10 })
+        eq("external window cannot have 'win'", pcall_err(api.nvim_win_set_config, 0, { win = 0 }))
+        -- OK to include "win" if external window is also reconfigured to a normal float.
```
`Currently accepts no other positioning configuration together with this.` is already mentioned for `external`; maybe already enough?

justinmk commented on src/nvim/winfloat.c:
```diff
@@ -38,15 +38,14 @@
 /// Create a new float.
 ///
 /// @param wp      if NULL, allocate a new window, otherwise turn existing window into a float.
```
was just a guess
